### PR TITLE
Increase max rate limit for crier controller rate limiter

### PR DESCRIPTION
**Pull request by timwangmusic**
We observe cases where k8s API controller throttling requests from crier controller. 

Previously we were using the `workqueue.DefaultControllerRateLimiter()`, which is calling `workqueue.NewMaxOfRateLimiter` internally with rate.Limit set to 10 and bucket size set to 100. And the setting for `NewItemExponentialFailureRateLimiter` is 5ms for base delay and 1000 seconds for max delay.

In this change, we double the QPS and bucket size and decrease max delay by half to 500 seconds and keep the base delay intact.

/cc @listx

### CODE REVIEW

**Comment by listx** [comment]
> Increase overall QPS and bucket size by 10 times
> Reduce per item max delay by half.

The idea (adjusting the rate limit settings) sounds fine, but I am not so sure about a 10x increase in something. Also, it's not clear from the diff what the current settings are. IOW how do we know that this won't negatively cause service disruption?

It would be safer to understand what the current settings are and to encode that into a PR. Then in another PR you could adjust those settings so that it's easy to see exactly how we are tweaking these settings.

Also you are missing an explanation of what these QPS and bucket sizes actually mean. Please add an explanation in your commit message and update the PR description accordingly. Thanks.

File: prow/crier/controller.go [hunk 1]
```diff
@@ -159,14 +163,14 @@ func (r *reconciler) reconcile(ctx context.Context, log *logrus.Entry, req recon
...
-			// The error above is alreay logged, so it would be duplicated
+			// The error above is already logged, so it would be duplicated
 			// effort to combine all errors to return, only capture the last
 			// error should be sufficient.
 			lastErr = err
 		}
 	}
 
-	if pj.Status.CompletionTime != nil {
+	if pj.Complete() {
```
Comment (listx) [inline prow/crier/controller.go:173]: This is not adjusting the rate limit. Please explain why you need to do this in the commit message.

File: prow/crier/controller.go [hunk 2]
```diff
@@ -159,14 +163,14 @@ func (r *reconciler) reconcile(ctx context.Context, log *logrus.Entry, req recon
 	for _, pjob := range pjs {
 		if err := criercommonlib.UpdateReportStateWithRetries(ctx, pjob, log, r.pjclientset, r.reporter.GetName()); err != nil {
 			log.WithError(err).Error("Failed to update report state on prowjob")
-			// The error above is alreay logged, so it would be duplicated
+			// The error above is already logged, so it would be duplicated
```
Comment (listx) [inline prow/crier/controller.go:166]: Please do not mix unrelated typo fixes into a single commit (you can do it in a separate commit).

File: prow/crier/controller.go [hunk 3]
```diff
@@ -68,7 +69,10 @@ func New(
 		Named(fmt.Sprintf("crier_%s", reporter.GetName())).
 		For(&prowv1.ProwJob{}).
 		WithOptions(controller.Options{MaxConcurrentReconciles: numWorkers,
-			RateLimiter: workqueue.DefaultControllerRateLimiter()}).
+			RateLimiter: workqueue.NewMaxOfRateLimiter(
+				workqueue.NewItemExponentialFailureRateLimiter(5*time.Millisecond, 500*time.Second),
+				// 100 qps, 1000 bucket size.
+				&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(100), 1000)})}).
```
Comment (listx) [inline prow/crier/controller.go:75]: How do we know these are safe settings?

Also, what settings are we currently using with workqueue.DefaultControllerRateLimiter()?

**Review by listx — COMMENTED** [review]
Please add more explanations to make it easier for reviewers to review. Thanks.

File: prow/crier/controller.go (hunk 2 above)
Comment (timwangmusic) [reply prow/crier/controller.go]: Reverted the change.

File: prow/crier/controller.go (hunk 1 above)
Comment (timwangmusic) [reply prow/crier/controller.go]: Reverted the change.

File: prow/crier/controller.go (hunk 3 above)
Comment (timwangmusic) [reply prow/crier/controller.go:75]: The `workqueue.DefaultControllerRateLimiter()` is calling `workqueue.NewMaxOfRateLimiter` internally with `rate.Limit` set to 10 and bucket size set to 100. I cannot say for sure that these are safe for sure, but we have example in the codebase that is using a higher `rate.Limit` and bucket size.

Comment (listx) [reply prow/crier/controller.go:75]: > The workqueue.DefaultControllerRateLimiter() is calling workqueue.NewMaxOfRateLimiter internally with rate.Limit set to 10 and bucket size set to 100.

This sort of background information is exactly what should be going into the commit message or PR description.

If you're not feeling comfortable with the size of the increase, then it would be prudent to make the bump less drastic. For example, a 20% increase, followed by another 20% increase, etc. So you'd have to break this PR up into several smaller ones (and then measure the impact after each one after it's rolled out).

File: prow/crier/controller.go (hunk 1 above)
Comment (listx) [reply prow/crier/controller.go]: > Reverted the change.

OK, but also for the future it would help reviewers if you explain why you did this in the first place (this is what I asked for...), just in case the reviewers (including myself) are missing some context. That sort of transparency in communication would help to create better PRs in general, certainly.

File: prow/crier/controller.go (hunk 2 above)
Comment (listx) [reply prow/crier/controller.go]: I would have preferred if you still fixed the (obviously wrong) typo as a separate commit. But as it is not the main focus of this PR, I'm OK with dropping it.

File: prow/crier/controller.go (hunk 3 above)
Comment (timwangmusic) [reply prow/crier/controller.go:75]: I would say double up the QPS and bucket size instead of 20% increase, wdyt?

Comment (listx) [reply prow/crier/controller.go:75]: SGTM, thanks.

Comment (timwangmusic) [reply prow/crier/controller.go]: Updated and also included the background info in commit message. Thanks.

File: prow/crier/controller.go (hunk 2 above)
Comment (timwangmusic) [reply prow/crier/controller.go:166]: Pushed a new commit with the typo fix.

**Comment by listx** [comment]
> Previously we were using the workqueue.DefaultControllerRateLimiter(), which is calling workqueue.NewMaxOfRateLimiter internally with rate.Limit set to 10 and bucket size set to 100.

Nit: Git commit messages shouldn't get too long on one line, so you should fix your editor to insert line breaks after every 70 or 80 columns.

> In this change, we double the QPS and bucket size

Makes sense.

> and decrease maximum delay by half.

Is this referring to the

    5*time.Millisecond, 500*time.Second)

part? So is the original setting 10ms and 1000 seconds? If so you should mention this too in the commit message. But also just like "bucket size" and "QPS", it would be good to understand how theses parameters are named (are both of them really just named "maximum delay" and "maximum delay"? probably not right? so please call out their names as they are, thanks).

As a courtesy to our future selves, please update the PR description to match any improvements to the commit message. This way your PR description will be up to date (and won't conflict, possibly, with what actually happened over the code review process).

**Comment by cjwagner** [comment]
Rather than increasing the quota (which will increase load on the API server even more), can we instead fix the retry behavior?
This doesn't address the underlying problem, it just mitigates it, and the additional load could be a problem.
